### ISSUE TITLE
New deferred path expansion logic greatly reduced io performance

### ISSUE THREAD

**Issue by kszlim** [pola-rs/polars#17584]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
# None, because it's doing some complicated queries on a huge dataset, but in terms of pseudocode it's something like:
from itertools import batched # This comes from python3.12
all_ids = ... # all the ids that i want to run work over, these are natural keys for my dataset to be partitioned on (they're hive partition values)
ldf = pl.scan_parquet(SOME_HUGE_HIVE_PARTITIONED_DIRECTORY, hive_partitioning=True)

batch_size = ...

for batch in batched(all_ids, batch_size):
    ldf = ldf.filter(pl.col("id_col").is_in(batch))
    ldf.collect() # but imagine this done in parallel in a separate ray actor
```

### Issue description

My code performs dramatically worse after the deferral of path expansion into collect.

I think the reason is because I'm using polars in a manner where I re-run my code with a cached lazyframe that gets filtered in a loop. (The filter is to batch up the work for [ray](https://docs.ray.io/en/latest/index.html)). The problem is because the `collect` now contains all the scanning work (I have a hive partitioned bucket with ~10k files), this gets repeated multiple times (whereas I previously was caching the path expansion work since my lazyframe was cached).

But even if I move my code outside of ray, and just try running it, this seems to dramatically reduce performance. I think my CPUs end up starved due to much lower network IO performance for path scanning.

Would it be possible to add an option to re-enable the eager path expansion into `collect`, ie. `eager_expand_paths: bool` or even revert this behavior?

### Expected behavior

Should be as fast as before this change. I'm still not sure why moving more correctness checks into `collect` time is desirable, unless there is a way to detach the source from the logical computational plan? Maybe being able to manipulate the source and the computational graph separately would be a nice evolution of the api?

### Installed versions

<details>

```
https://github.com/pola-rs/polars/commit/64b45a89813ceb25532d2678f9e9d5b85659b35e
```

</details>

**Comment by ritchie46** [comment]
It has to go to IR though. It was wrong before. We can come up with a caching mechanism same as with schema. 

>  I'm still not sure why moving more correctness checks into collect 

Because we want to be able to run Polars queries server side. A lazyframe is a DSL, it should not execute a query yet. Think of it as being the SQL that you send to the database.

**Comment by ritchie46** [comment]
We can keep best of both worlds, we can put the expanded paths back the DSL.

**Comment by mcrumiller** [comment]
@ritchie46 is there a way to build a DSL and then attach it later to a source?
